torch/onnx/_internal/fx/registration.py

```diff
@@ -192,83 +80,136 @@ def _initiate_registry_from_torchlib(
+                symbolic_function = SymbolicFunction(
+                    onnx_function=overload_func,
+                    op_full_name=internal_name,
+                    is_custom=False,
                 )
+                self._register(symbolic_function)
 
     @_beartype.beartype
-    def register(
-        self, name: str, opset: OpsetVersion, func: Callable, custom: bool = True
-    ) -> None:
-        """Registers overloaded functions to an ATen operator (overload).
+    def _register(self, symbolic_function: SymbolicFunction) -> None:
+        """Registers a SymbolicFunction to an operator.
 
         Args:
-            name: The qualified name of the function to register. In the form of 'domain::op'.
-                E.g. 'aten::add' or 'aten::pow.int'.
-            opset: The opset version of the function to register.
-            func: The symbolic function to register.
-            custom: Whether the function is a custom function that overrides existing ones.
-
-        Raises:
-            ValueError: If the separator '::' is not in the name.
+            symbolic_function: The SymbolicFunction to register.
         """
-        if "::" not in name:
-            raise ValueError(
-                f"The name must be in the form of 'domain::op', not '{name}'"
-            )
-        symbolic_functions = self._registry.setdefault(
-            name, _SymbolicFunctionGroup(name)
-        )
-        if custom:
-            symbolic_functions.add_custom(func, opset)
-        else:
-            symbolic_functions.add(func, opset)
+        self._registry[symbolic_function.op_full_name].append(symbolic_function)
 
     @_beartype.beartype
-    def unregister(self, name: str, opset: OpsetVersion) -> None:
-        """Unregisters overloaded functions to an ATen operator (overload).
+    def register_custom_op(
+        self,
+        function: Union["onnxscript.OnnxFunction", "onnxscript.TracedOnnxFunction"],
+        domain: str,
+        op_name: str,
+        overload: Optional[str] = None,
+    ) -> None:
+        """Registers a custom operator: torch.ops.<domain>.<op_name>.<overload>.
 
         Args:
-            name: The qualified name of the function to unregister.
-            opset: The opset version of the function to unregister.
+            function: The onnx-sctip function to register.
+            domain: The domain of the operator to register.
+            op_name: The name of the operator to register.
+            overload: The overload of the operator to register. If it's default overload,
+                leave it to None.
+
+        Raises:
+            ValueError: If the name is not in the form of 'domain::op'.
         """
-        if name not in self._registry:
-            return
-        self._registry[name].remove_custom(opset)
+        internal_name = OpName.from_string(
+            domain=domain, op_name=op_name, overload=overload
+        ).internal_name()
```

Comment:
What does "internal" refer to? Internal to PyTorch, or internal to torch.onnx? Does it matter? Should we call it full_qualified_name or something more descriptive?

Reply:
Like internal usage for converter team (Baasically, we normalize name first when we get the op name.). I think it will be used a lot, so didn't want to name it too long. Any suggestion?

Reply:
How about just str(OpName())? Is there any other “names” than internal name?

Reply:
You mean like writing __str__? There is overloadpacket name style, and overload name style. Basically, they would even possibly take aten::add and aten::add.default the same. The function is trying to normalize the name and give us a common internal name for our own usage.

Reply:
I see. So should str just be the normalized string we use internally?

Reply:
Yes. So wherever you get op name like stuff (ex: from opoverload/opoverloadpacket attribute, or self create namespace/op_name/overload, ...) Put it into OpName and use internal_name() instead.

Reply:
Not sure if I'm the only one here but we could also just add `__hash__` to `OpName` and use `OpName` as key for registry.

Reply:
> Not sure if I'm the only one here but we could also just add `__hash__` to `OpName` and use `OpName` as key for registry.

I like this too. Frozen=True and eq=True makes dataclass generate a hash method: https://docs.python.org/3/library/dataclasses.html

Initially I thought this would be inefficient because I was thinking about eq and forgot about hash completely.

Reply:
Done